### PR TITLE
net-mail/dot-forward: fix ebuild to work with clang16

### PR DESCRIPTION
**Pull request by Schievel1**
Closes: https://bugs.gentoo.org/870856
Signed-off-by: Pascal Jäger <pascal.jaeger@leimstift.de>

### CODE REVIEW

**Comment by thesamesam** [comment]
I get a tonne of errors still with this, as we still have to fix implicit func. decls and such with gnu89, we just don't have to worry about prototypes:
```
make: *** [Makefile:266: seek_set.o] Error 1
dot-forward.c: In function ‘run’:
dot-forward.c:78:7: error: implicit declaration of function ‘pipe’ [-Werror=implicit-function-declaration]
   78 |   if (pipe(pi) == -1)
      |       ^~~~
dot-forward.c:81:19: error: implicit declaration of function ‘fork’ [-Werror=implicit-function-declaration]
   81 |   switch (child = fork()) {
      |                   ^~~~
dot-forward.c:85:7: error: implicit declaration of function ‘close’ [-Werror=implicit-function-declaration]
   85 |       close(pi[1]);
      |       ^~~~~
dot-forward.c:86:11: error: implicit declaration of function ‘fd_move’ [-Werror=implicit-function-declaration]
   86 |       if (fd_move(0,pi[0]) == -1)
      |           ^~~~~~~
dot-forward.c:89:7: error: implicit declaration of function ‘sig_pipedefault’ [-Werror=implicit-function-declaration]
   89 |       sig_pipedefault();
      |       ^~~~~~~~~~~~~~~
dot-forward.c:90:7: error: implicit declaration of function ‘execv’ [-Werror=implicit-function-declaration]
   90 |       execv(*args,args);
      |       ^~~~~
dot-forward.c: In function ‘readcontrols’:
dot-forward.c:141:7: error: implicit declaration of function ‘chdir’ [-Werror=implicit-function-declaration]
  141 |   if (chdir(auto_qmail) == -1)
      |       ^~~~~
dot-forward.c:160:7: error: implicit declaration of function ‘fchdir’ [-Werror=implicit-function-declaration]
  160 |   if (fchdir(fddir) == -1)
      |       ^~~~~~
dot-forward.c: In function ‘gotaddr’:
dot-forward.c:189:12: error: implicit declaration of function ‘case_diffb’ [-Werror=implicit-function-declaration]
  189 |       if (!case_diffb(address.s,address.len,user)) {
      |            ^~~~~~~~~~
dot-forward.c: In function ‘main’:
dot-forward.c:389:3: error: implicit declaration of function ‘sig_pipeignore’ [-Werror=implicit-function-declaration]
  389 |   sig_pipeignore();
      |   ^~~~~~~~~~~~~~
cc1: some warnings being treated as errors
make: *** [Makefile:95: dot-forward.o] Error 1
rm -f trywaitp.o trywaitp
 * ERROR: net-mail/dot-forward-0.71-r4::gentoo failed (compile phase):
 *   emake failed
```

**Comment by Schievel1** [comment]
> I get a tonne of errors still with this, as we still have to fix implicit func. decls and such with gnu89, we just don't have to worry about prototypes:
> 


Think it works now.

File: net-mail/dot-forward/dot-forward-0.71-r4.ebuild [hunk 1]
```diff
@@ -46,6 +48,7 @@ src_prepare() {
 }
 
 src_configure() {
+	append-cflags
```
Comment (thesamesam) [inline net-mail/dot-forward/dot-forward-0.71-r4.ebuild:51]: Did you mean to put -std=gnu89 or something?

Comment (Schievel1) [reply net-mail/dot-forward/dot-forward-0.71-r4.ebuild:51]: Head -> table. 
i actually meant to remove this line altogether.

File: net-mail/dot-forward/files/0.71-fix-build-for-clang16.patch [hunk 2]
```diff
@@ -0,0 +1,1826 @@
...
+--- a/Makefile
++++ b/Makefile
+@@ -19,8 +19,8 @@ compile alloc_re.c alloc.h byte.h
+ auto-ccld.sh: \
+ conf-cc conf-ld warn-auto.sh
+ 	( cat warn-auto.sh; \
+-	echo CC=\'`head -1 conf-cc`\'; \
+-	echo LD=\'`head -1 conf-ld`\' \
++	echo CC=\'`head -n 1 conf-cc`\'; \
++	echo LD=\'`head -n 1 conf-ld`\' \
```
Comment (thesamesam) [inline net-mail/dot-forward/files/0.71-fix-build-for-clang16.patch:12]: nit: the head changes here are actually from the eclass thing we call in the ebuild. I'd reapply the patch with `git add -p` so we don't include this, or maybe split this part into a separate patch and drop the use in the ebuild. Up to you.

It's not terrible to have it mixed here, but it's not ideal, as it's not related.

File: net-mail/dot-forward/dot-forward-0.71-r4.ebuild [hunk 3]
```diff
@@ -1,9 +1,9 @@
-# Copyright 1999-2021 Gentoo Authors
+# Copyright 1999-2022 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
-EAPI=7
+EAPI=8
 
-inherit fixheadtails qmail
+inherit fixheadtails flag-o-matic qmail
```
Comment (thesamesam) [inline net-mail/dot-forward/dot-forward-0.71-r4.ebuild:6]: Missing revbump (for the implicit func. decl stuff, we don't want to take any risks).

But also, while we don't normally revbump for "build time fixes", the nature of a build-time fix is kind of varied. a huge patch like this merits a revbump anyway (even if there were no expected possible change in runtime behaviour) just because there's a non-zero chance it'll affect behaviour or cause a bug we want to know about sooner rather than later.

File: net-mail/dot-forward/dot-forward-0.71-r5.ebuild [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+inherit fixheadtails qmail
+
+DESCRIPTION="reads sendmail's .forward files under qmail"
+HOMEPAGE="http://cr.yp.to/dot-forward.html"
+SRC_URI="http://cr.yp.to/software/${P}.tar.gz"
+
+LICENSE="public-domain"
+SLOT="0"
+KEYWORDS="~alpha amd64 arm ~hppa ~ia64 ~m68k ~mips ppc ppc64 ~s390 sparc x86"
```
Comment (thesamesam) [inline net-mail/dot-forward/dot-forward-0.71-r5.ebuild:14]: You've kept stable keywords. Either git mv in this case or copy and ekeyword ~all.

File: net-mail/dot-forward/files/0.71-fix-build-for-clang16.patch [hunk 5]
```diff
@@ -0,0 +1,1662 @@
...
+
+Signed-off-by: Pascal Jäger <pascal.jaeger@leimstift.de>
+
+--- a/alloc.c
++++ b/alloc.c
+@@ -1,6 +1,6 @@
+ #include "alloc.h"
+ #include "error.h"
+-extern char *malloc();
++extern char *malloc(int n);
```
Comment (thesamesam) [inline net-mail/dot-forward/files/0.71-fix-build-for-clang16.patch:15]: fwiw for these, I try to drop them if I can, but IIRC it was a pain here

**Comment by thesamesam** [comment]
I get some warnings which I'll fix on merge:
```
dot-forward.c:382:1: warning: return type of 'main' is not 'int' [-Wmain-return-type]
void main(int argc,char **argv)
^
dot-forward.c:382:1: note: change return type to 'int'
void main(int argc,char **argv)
^~~~
int
1 warning generated.
rm -f tryvfork.o tryvfork
./compile qmail.c
./load auto-str substdio.a error.a str.a
alloc.c:3:14: warning: incompatible redeclaration of library function 'malloc' [-Wincompatible-library-redeclaration]
extern char *malloc(int n);
             ^
alloc.c:3:14: note: 'malloc' is a builtin with type 'void *(unsigned long)'
```

Non-fatal, but still.
